15794 Make "related objects" dynamic

alehaa opened the pull request:
<!--
    Thank you for your interest in contributing to NetBox! Please note that
    our contribution policy requires that a feature request or bug report be
    approved and assigned prior to opening a pull request. This helps avoid
    waste time and effort on a proposed change that we might not be able to
    accept.

    IF YOUR PULL REQUEST DOES NOT REFERENCE AN ISSUE WHICH HAS BEEN ASSIGNED
    TO YOU, IT WILL BE CLOSED AUTOMATICALLY.

    Please specify your assigned issue number on the line below.
-->
### Fixes: #15794

<!--
    Please include a summary of the proposed changes below.
-->

Instead of hardcoding relationships between models for the detail view, they are now dynamically generated.

#### Known limitations

1. If a related model doesn't have a `list` view, rendering of the [related objects pane](https://github.com/netbox-community/netbox/blob/85db007ff585370b03b2d88d8f43e6fa9b94bc8e/netbox/templates/inc/panels/related_objects.html#L8) will fail. This limitation was already present before, but related models listed were explicitly chosen to do so. For all views shipped with NetBox I've added a filter if needed (`omit` parameter). However, external plugins might trigger this unintentionally, when named view is not available, but the relationship has not been setup with `related_name='+'` parameter. Switching to `verified_view` in the panel's template should resolve this, but might decrease performance slightly.

2. If a model introduces several relations to another model, they will all be listed with the same model name but different URL query parameters. This is not the case for models included in NetBox. As before, plugin developers can still use an extra panel for this special purpose and extra context about the different types of relationships.

alehaa committed:
Closes #15794: Make "related objects" dynamic

Instead of hardcoding relationships between models for the detail view,
they are now dynamically generated.

alehaa committed:
Fix related models call

jeremystretch commented on netbox/utilities/views.py:
```diff
@@ -140,6 +144,51 @@ def get_return_url(self, request, obj=None):
+    Provides logic for collecting all related models for the currently viewed model.
+    """
+
+    def get_extra_related_models(self, request, instance):
```
What is the purpose of this method? It seems like maybe a hook for monkey patching, which IMO should be discouraged. If needed, we could potentially extend the underlying `get_related_models()` utility function (not the class method) to collect additional types of relationships.

Or, if it's intended to be overridden naturally, we could introduce an `extra` keyword argument to the `get_related_models()` method to pass these.

alehaa replied:
You are right, its intended to be used as a hook by view classes, e.g. when not only direct relations should be collected.

Although I prefer a generic approach that implements some logic in the utility function, I don't know how we should determine which transient relations are interesting for object views. So for now, I'd remove this method and add an `extra` keyword argument to the class method instead.

jeremystretch committed:
Merge branch 'develop' into feature/15794-related-objects

alehaa committed:
Remove extra related models hook

Instead of providing a rarely used hook method, additional related
models can now be passed directly to the lookup method.

alehaa committed:
Merge branch 'develop' into feature/15794-related-objects

jeremystretch commented on netbox/circuits/views.py:
```diff
@@ -156,19 +147,22 @@ class ProviderNetworkListView(generic.ObjectListView):
+            'related_models': self.get_related_models(
+                request,
+                instance,
+                [CircuitTermination],
```
We should be able to include circuit terminations now that #15496 has been completed.

jeremystretch commented on netbox/dcim/views.py:
```diff
@@ -380,31 +386,23 @@ class SiteListView(generic.ObjectListView):
+                request,
+                instance,
+                [CableTermination, CircuitTermination],
+                (
```
Looks like we need to pull in ASNs too.

jeremystretch commented on netbox/dcim/views.py:
```diff
@@ -912,16 +889,16 @@ class DeviceTypeListView(generic.ObjectListView):
-
         return {
-            'related_models': related_models,
+            'related_models': self.get_related_models(request, instance, [
```
Including the keyword improves clarity.
```suggestion
            'related_models': self.get_related_models(request, instance, omit=[
```

jeremystretch commented on netbox/dcim/views.py:
```diff
@@ -1151,16 +1128,16 @@ class ModuleTypeListView(generic.ObjectListView):
-
         return {
-            'related_models': related_models,
+            'related_models': self.get_related_models(request, instance, [
```
```suggestion
            'related_models': self.get_related_models(request, instance, omit=[
```

jeremystretch commented on netbox/ipam/views.py:
```diff
@@ -52,7 +47,7 @@ def get_extra_context(self, request, instance):
 
         return {
-            'related_models': related_models,
+            'related_models': self.get_related_models(request, instance),
```
This returns two entries for "Interfaces," because we're querying both device and VM interfaces. The simplest solution would be to exclude both, which IMO is reasonable.

jeremystretch commented on netbox/ipam/views.py:
```diff
@@ -272,17 +263,12 @@ class ASNListView(generic.ObjectListView):
-
         return {
-            'related_models': related_models,
+            'related_models': self.get_related_models(request, [instance]),
```
Site and Provider need to be pulled explicitly as they are both M2M relationships.

alehaa committed:
Merge branch 'develop' into feature/15794-related-objects

alehaa committed:
Fix relations view for ASNs

ASNs have ManyToMany relationships and therefore can't used automatic
resolving. Explicit relations have been restored as before.

alehaa committed:
Add method call keywords for clarification

alehaa committed:
Cleanup related models

jeremystretch reviewed (approved):
Thanks @alehaa!
